# Usage Example

In [1]:
from scMFF_pipeline import *
dir = 'Example_data/'
model_name = 'KNN'

scMFF_model(dir, model_name)

---------------------------------------------------------------
Step 0: Data loading in progress...
Data loading complete.
---------------------------------------------------------------
Step 1: Data preprocessing in progress...


/home/sunnan/anaconda3/envs/SunNan_env/lib/python3.10/site-packages/anndata/_core/anndata.py:1113: DeprecationWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, pd.CategoricalDtype) instead
  if not is_categorical_dtype(df_full[k]):
/mnt/202001Main/BioData/Transcriptome_classification/scMFF/scMFF/Preprocessing.py:8: DeprecationWarning: `log1p` is deprecated since scVelo v0.3.0 and will be removed in a future version. Please use `log1p` from `scanpy.pp` instead.
  scv.pp.log1p(adata)


Extracted 2000 highly variable genes.
Data preprocessing complete.
---------------------------------------------------------------
Step 2: Feature extraction in progress...
(1) Statistics-based Feature: expr_2000.csv 	 (2998, 2000)
(2) Information-based Feature: scPSD_embedding.txt.gz 	 (2998, 2000)
(3) Matrix Factorization-based Feature: PCA_embedding.txt.gz 	 (2998, 266)
Step1: Start filter and generating CSV
Input scRNA data in CSV format is validated, start reading...
Data loaded, start filtering...
After preprocessing, 14118 genes remaining
After preprocessing, 2998 cells have 0.99 nonzero
Preprocessing Done. Total Running Time: 10.907299757003784 seconds


2025-05-15 15:16:41.514569: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


Using device:cuda
---0:00:00---scRNA starts loading.
Load expression in csv format
---0:00:02---scRNA has been successfully loaded
---0:00:02---TrainLoader has been successfully prepared.
---0:00:03---Pytorch model ready.
Start training...
Train Epoch: 1 [0/2998 (0%)]	Loss: 237.400246
====> Epoch: 1 Average loss: 237.4002
Train Epoch: 2 [0/2998 (0%)]	Loss: 198.652643
====> Epoch: 2 Average loss: 198.6526
Train Epoch: 3 [0/2998 (0%)]	Loss: 189.751564
====> Epoch: 3 Average loss: 189.7516
Train Epoch: 4 [0/2998 (0%)]	Loss: 149.167591
====> Epoch: 4 Average loss: 149.1676
Train Epoch: 5 [0/2998 (0%)]	Loss: 138.537483
====> Epoch: 5 Average loss: 138.5375
Train Epoch: 6 [0/2998 (0%)]	Loss: 128.193358
====> Epoch: 6 Average loss: 128.1934
Train Epoch: 7 [0/2998 (0%)]	Loss: 125.429661
====> Epoch: 7 Average loss: 125.4297
Train Epoch: 8 [0/2998 (0%)]	Loss: 120.841186
====> Epoch: 8 Average loss: 120.8412
Train Epoch: 9 [0/2998 (0%)]	Loss: 114.851484
====> Epoch: 9 Average loss: 114.8515
Trai

feature1  feature2  feature3  feature4  feature5  feature6  \
Accuracy   0.960309  0.979320  0.975985  0.974650  0.982990  0.758835   
Precision  0.962588  0.979444  0.976576  0.974951  0.983186  0.761053   
Recall     0.960309  0.979320  0.975985  0.974650  0.982990  0.758835   
F1         0.960076  0.979313  0.975896  0.974613  0.982959  0.753840   
MCC        0.949851  0.973477  0.969405  0.967539  0.978250  0.689785   

           feature7  feature8  feature9  feature10  
Accuracy   0.955302  0.980322  0.977319   0.979654  
Precision  0.956969  0.980684  0.977808   0.979905  
Recall     0.955302  0.980322  0.977319   0.979654  
F1         0.955218  0.980296  0.977275   0.979609  
MCC        0.943005  0.974869  0.971040   0.973980

### Feature Embedding Files and Their Dimensions

| Feature ID | File Name                                                               | Dimension      |
|------------|-------------------------------------------------------------------------|----------------|
| Feature0   | expr.csv                                                                | (33694, 2998)  |
| Feature1   | expr_2000.csv                                                           | (2000, 2998)   |
| Feature2   | scPSD_embedding.txt.gz                                                  | (2998, 2000)   |
| Feature3   | PCA_embedding.txt.gz                                                    | (2998, 266)    |
| Feature4   | scGNN_embedding.csv                                                     | (2998, 129)    |
| Feature5   | Fusion_sum-expr-scpsd-scgnn-pca.txt.gz                                  | (2998, 2000)   |
| Feature6   | Fusion_hadamard-expr-scpsd-scgnn-pca.txt.gz                             | (2998, 2000)   |
| Feature7   | Fusion_attention-expr-scpsd-scgnn-pca.txt.gz                            | (2998, 2000)   |
| Feature8   | Fusion_moe-expr-scpsd-scgnn-pca.txt.gz                                  | (2998, 128)    |
| Feature9   | Fusion_residual-expr-scpsd-scgnn-pca.txt.gz                             | (2998, 128)    |
| Feature10  | Fusion_transformer-expr-scpsd-scgnn-pca.txt.gz                          | (2998, 128)    |



# Step-by-step testing scMFF_pipeline.py

In [1]:
pwd

'/mnt/202001Main/BioData/Transcriptome_classification/scMFF/scMFF'

In [2]:
ls

Classifier.py           Model.png         README.md          utils.py
Example_data/           Preprocessing.py  scGNN/
Example_data_complete/  psdPy/            scMFF_pipeline.py
Fusion.py               __pycache__/      test.ipynb


In [3]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
from Preprocessing import *
from Fusion import *
from Classifier import *
from utils import *

import sys  
import os
from sklearn.decomposition import PCA

In [4]:
dir = 'Example_data/'
model_name = 'KNN'

In [5]:
print('---------------------------------------------------------------')
print('Step 0: Data loading in progress...')
df = pd.read_csv(dir + 'expr.csv', index_col=0, header=0)

CellType = pd.read_csv(dir + 'cluster.csv', header=None).iloc[:, 0].values
label_encoder = LabelEncoder()
Y = label_encoder.fit_transform(CellType)
num_classes = len(np.unique(Y))

print('Data loading complete.')  

---------------------------------------------------------------
Step 0: Data loading in progress...
Data loading complete.


In [6]:
####        Step 1: Data Pre-processing          ########################################################
print('---------------------------------------------------------------')
print('Step 1: Data preprocessing in progress...')
df0 = preprocess_RNASeq(df)
df0.to_csv(dir + 'expr_2000.csv')
print('Data preprocessing complete.')

---------------------------------------------------------------
Step 1: Data preprocessing in progress...


/home/sunnan/anaconda3/envs/SunNan_env/lib/python3.10/site-packages/anndata/_core/anndata.py:1113: DeprecationWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, pd.CategoricalDtype) instead
  if not is_categorical_dtype(df_full[k]):
/mnt/202001Main/BioData/Transcriptome_classification/scMFF/scMFF/Preprocessing.py:8: DeprecationWarning: `log1p` is deprecated since scVelo v0.3.0 and will be removed in a future version. Please use `log1p` from `scanpy.pp` instead.
  scv.pp.log1p(adata)


Extracted 2000 highly variable genes.
Data preprocessing complete.


In [7]:
print('---------------------------------------------------------------')
print('Step 2: Feature extraction in progress...')
    
# a. Statistics-based Feature: expr_2000.csv
feature1 = df0.T.values
print('(1) Statistics-based Feature: expr_2000.csv \t', feature1.shape)
    

---------------------------------------------------------------
Step 2: Feature extraction in progress...
(1) Statistics-based Feature: expr_2000.csv 	 (2998, 2000)


In [8]:
# b. Information-based Feature: scPSD      
sys.path.append('scMFF/psdPy') 
from psdPy import psd
X_scPSD = psd(df0.T) 
np.savetxt(dir + 'scPSD_embedding.txt.gz', X_scPSD)
feature2 = X_scPSD
print('(2) Information-based Feature: scPSD_embedding.txt.gz \t', feature2.shape)
    

(2) Information-based Feature: scPSD_embedding.txt.gz 	 (2998, 2000)


In [9]:
# c. Matrix Factorization-based Feature: PCA
pca = PCA(n_components=266) 
X_PCA = pca.fit_transform(df0.T)
np.savetxt(dir + 'PCA_embedding.txt.gz', X_PCA)
feature3 = X_PCA
print('(3) Matrix Factorization-based Feature: PCA_embedding.txt.gz \t', feature3.shape)
    

(3) Matrix Factorization-based Feature: PCA_embedding.txt.gz 	 (2998, 266)


In [10]:
sys.path.append('scMFF') 
dir_path1 = 'Example_data/'
preprocess_command = f"python -W ignore scGNN/PreprocessingscGNN.py --datasetName expr.csv --datasetDir {dir_path1} --LTMGDir {dir_path1} --filetype CSV --geneSelectnum 2000"    
os.system(preprocess_command)

Step1: Start filter and generating CSV
Input scRNA data in CSV format is validated, start reading...
Data loaded, start filtering...
After preprocessing, 14118 genes remaining
After preprocessing, 2998 cells have 0.99 nonzero
Preprocessing Done. Total Running Time: 10.51564908027649 seconds


0

In [11]:
ls

Classifier.py           Model.png         README.md          utils.py
Example_data/           Preprocessing.py  scGNN/
Example_data_complete/  psdPy/            scMFF_pipeline.py
Fusion.py               __pycache__/      test.ipynb


In [12]:
# d. Deep Learningg-based Feature: scGNN
sys.path.append('scMFF/') 
dir_path1 = '/home/sunnan/BioData/Transcriptome_classification/scMFF/scMFF/Example_data/'
preprocess_command = f"python -W ignore scGNN/PreprocessingscGNN.py --datasetName expr.csv --datasetDir {dir_path1} --LTMGDir {dir_path1} --filetype CSV --geneSelectnum 2000"    
os.system(preprocess_command)

dir_path2 = '/home/sunnan/BioData/Transcriptome_classification/scMFF/scMFF/'
scgnn_command = f"python -W ignore scGNN/scGNN.py --datasetName Example_data --datasetDir {dir_path2} --outputDir {dir_path1} --EM-iteration 2 --Regu-epochs 50 --EM-epochs 20 --quickmode --nonsparseMode"
os.system(scgnn_command)
feature4 = pd.read_csv(dir + 'scGNN_embedding.csv', index_col=0, header=0).values
print('(4) Deep Learningg-based Feature: scGNN_embedding.csv \t', feature4.shape)


Step1: Start filter and generating CSV
Input scRNA data in CSV format is validated, start reading...
Data loaded, start filtering...
After preprocessing, 14118 genes remaining
After preprocessing, 2998 cells have 0.99 nonzero
Preprocessing Done. Total Running Time: 10.91727066040039 seconds


2025-05-15 15:07:32.561401: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


Using device:cuda
---0:00:00---scRNA starts loading.
Load expression in csv format
---0:00:01---scRNA has been successfully loaded
---0:00:01---TrainLoader has been successfully prepared.
---0:00:02---Pytorch model ready.
Start training...
Train Epoch: 1 [0/2998 (0%)]	Loss: 237.400246
====> Epoch: 1 Average loss: 237.4002
Train Epoch: 2 [0/2998 (0%)]	Loss: 198.652643
====> Epoch: 2 Average loss: 198.6526
Train Epoch: 3 [0/2998 (0%)]	Loss: 189.751564
====> Epoch: 3 Average loss: 189.7516
Train Epoch: 4 [0/2998 (0%)]	Loss: 149.167591
====> Epoch: 4 Average loss: 149.1676
Train Epoch: 5 [0/2998 (0%)]	Loss: 138.537483
====> Epoch: 5 Average loss: 138.5375
Train Epoch: 6 [0/2998 (0%)]	Loss: 128.193358
====> Epoch: 6 Average loss: 128.1934
Train Epoch: 7 [0/2998 (0%)]	Loss: 125.429661
====> Epoch: 7 Average loss: 125.4297
Train Epoch: 8 [0/2998 (0%)]	Loss: 120.841186
====> Epoch: 8 Average loss: 120.8412
Train Epoch: 9 [0/2998 (0%)]	Loss: 114.851484
====> Epoch: 9 Average loss: 114.8515
Trai

In [13]:
print('Feature extraction complete.')

Feature extraction complete.


In [14]:
####        Step 3: Cross Fusion Mudule          ########################################################
print('---------------------------------------------------------------')
print('Step 3: Cross Fusion in progress...')
[n_sample, n_feature] = feature1.shape
Fusion_sum, Fusion_hadamard, Fusion_attention = [], [], []
Fused_moe, Fused_residual, Fused_transformer = [], [], []

i = 0
max_dim = max(len(feature1[i, :]), len(feature2[i, :]), len(feature3[i, :]), len(feature4[i, :]))
output_dim = min(len(feature1[i, :]), len(feature2[i, :]), len(feature3[i, :]), len(feature4[i, :]))
   
moefusion = MoEFusion(max_dim * 4, output_dim)
residualfusion = ResidualFusion(max_dim, output_dim)
transformerfusion = TransformerConcatFusion(max_dim, output_dim)

for i in range(n_sample):
    print('The '+str(i+1)+'-th cell embedding')
    weights = [1/4, 1/4, 1/4, 1/4] 
    fusion_sum = weighted_sum_fusion([feature1[i,:], feature2[i,:], feature3[i,:], feature4[i,:]], weights)
    fusion_hadamard = hadamard_product_fusion([feature1[i,:], feature2[i,:], feature3[i,:], feature4[i,:]])
        
    features = [torch.tensor(feature1[i,:], dtype=torch.float32), torch.tensor(feature2[i,:], dtype=torch.float32), torch.tensor(feature3[i,:], dtype=torch.float32), torch.tensor(feature4[i,:], dtype=torch.float32)]
    labels = torch.tensor(Y[i])    
    AttentionFusion_model = train_AttentionFusion(features, labels, num_classes)
    fusion_attention,logits = AttentionFusion_model(features)

    Fusion_sum.append(fusion_sum)
    Fusion_hadamard.append(fusion_hadamard)
    Fusion_attention.append(fusion_attention.detach().cpu().numpy())

    f1 = zero_pad_feature(feature1[i, :],max_dim)
    f2 = zero_pad_feature(feature2[i, :],max_dim)
    f3 = zero_pad_feature(feature3[i, :],max_dim)
    f4 = zero_pad_feature(feature4[i, :],max_dim) 

    fused_moe = moefusion(f1, f2, f3, f4)
    fused_residual = residualfusion(f1, f2, f3, f4)
    fused_transformer = transformerfusion(f1, f2, f3, f4)  

    Fused_moe.append(fused_moe)
    Fused_residual.append(fused_residual)
    Fused_transformer.append(fused_transformer)



---------------------------------------------------------------
Step 3: Cross Fusion in progress...
The 1-th cell embedding
The 2-th cell embedding
The 3-th cell embedding
The 4-th cell embedding
The 5-th cell embedding
The 6-th cell embedding
The 7-th cell embedding
The 8-th cell embedding
The 9-th cell embedding
The 10-th cell embedding
The 11-th cell embedding
The 12-th cell embedding
The 13-th cell embedding
The 14-th cell embedding
The 15-th cell embedding
The 16-th cell embedding
The 17-th cell embedding
The 18-th cell embedding
The 19-th cell embedding
The 20-th cell embedding
The 21-th cell embedding
The 22-th cell embedding
The 23-th cell embedding
The 24-th cell embedding
The 25-th cell embedding
The 26-th cell embedding
The 27-th cell embedding
The 28-th cell embedding
The 29-th cell embedding
The 30-th cell embedding
The 31-th cell embedding
The 32-th cell embedding
The 33-th cell embedding
The 34-th cell embedding
The 35-th cell embedding
The 36-th cell embedding
The 37-th

In [15]:
np.savetxt(dir + 'Fusion_sum-expr-scpsd-scgnn-pca.txt.gz', Fusion_sum)
np.savetxt(dir + 'Fusion_hadamard-expr-scpsd-scgnn-pca.txt.gz', Fusion_hadamard)
np.savetxt(dir + 'Fusion_attention-expr-scpsd-scgnn-pca.txt.gz', Fusion_attention)

np.savetxt(dir + 'Fusion_moe-expr-scpsd-scgnn-pca.txt.gz', np.array([fused_moe.detach().numpy().flatten() for fused_moe in Fused_moe]))
np.savetxt(dir + 'Fusion_residual-expr-scpsd-scgnn-pca.txt.gz', np.array([fused_residual.detach().numpy().flatten() for fused_residual in Fused_residual]))
np.savetxt(dir + 'Fusion_transformer-expr-scpsd-scgnn-pca.txt.gz', np.array([fused_transformer.detach().numpy().flatten() for fused_transformer in Fused_transformer]))
    
feature5 = Fusion_sum
feature6 = Fusion_hadamard
feature7 = Fusion_attention
feature8 = np.array([fused_moe.detach().numpy().flatten() for fused_moe in Fused_moe])
feature9 = np.array([fused_residual.detach().numpy().flatten() for fused_residual in Fused_residual])
feature10 = np.array([fused_transformer.detach().numpy().flatten() for fused_transformer in Fused_transformer]) 


In [16]:
print('Cross Fusion complete.')

Cross Fusion complete.


In [17]:
####        Step 4: Classification Module        ########################################################
unique_classes, counts = np.unique(Y, return_counts=True)
min_count = 3
for cls in unique_classes:
    if counts[cls] < min_count:
        Y[Y == cls] = max(unique_classes) + 1 
new_label_encoder = LabelEncoder()
y = new_label_encoder.fit_transform(Y)

model = get_model(model_name)

In [18]:
results1 = cross_validation_model(feature1, y, model)
results2 = cross_validation_model(feature2, y, model)
results3 = cross_validation_model(feature3, y, model)
results4 = cross_validation_model(feature4, y, model)
results5 = cross_validation_model(feature5, y, model)
results6 = cross_validation_model(feature6, y, model)
results7 = cross_validation_model(feature7, y, model)
results8 = cross_validation_model(feature8, y, model)
results9 = cross_validation_model(feature9, y, model)
results10 = cross_validation_model(feature10, y, model)

In [19]:
df = pd.DataFrame({
        'feature1': results1,
        'feature2': results2,
        'feature3': results3,
        'feature4': results4,
        'feature5': results5,
        'feature6': results6,
        'feature7': results7,
        'feature8': results8,
        'feature9': results9,
        'feature10': results10
    })

df.to_excel(dir + 'results.xlsx')

In [20]:
df

feature1  feature2  feature3  feature4  feature5  feature6  \
Accuracy   0.960309  0.979320  0.975651  0.974650  0.982990  0.760502   
Precision  0.962588  0.979444  0.976313  0.974951  0.983186  0.762681   
Recall     0.960309  0.979320  0.975651  0.974650  0.982990  0.760502   
F1         0.960076  0.979313  0.975575  0.974613  0.982959  0.755224   
MCC        0.949851  0.973477  0.968990  0.967539  0.978250  0.691855   

           feature7  feature8  feature9  feature10  
Accuracy   0.959307  0.984324  0.975651   0.981655  
Precision  0.960774  0.984516  0.976109   0.981883  
Recall     0.959307  0.984324  0.975651   0.981655  
F1         0.959209  0.984300  0.975561   0.981628  
MCC        0.948148  0.979951  0.968935   0.976541